File: Array/Jump Game III/By Recursion.cpp
// Method - Recursion(Optimised)
// Time Complexity T - O(n)
// Sapce Complexity S - O(n), recursion take "n" space


class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
         
        // if start is greater than or equal to 0;
        // and start value is not more than the size of the arr
        // and value of start in arr is greater than or equal to 0
        // then do the following
        if (start >= 0 && start < arr.size() && arr[start] >= 0){
            
            // base case
            if (arr[start] == 0){
                return true;
            }   
            
            // converting the value of arr[start] to negative value
            arr[start] = -1 * arr[start];
            
            // recursion case
            // start+arr[start] -> i +arr[i]
            // start-arr[start] -> i -arr[i]
            // travese in both the way and return the value
            return canReach(arr, start+arr[start]) || canReach(arr, start-arr[start]); 
        }
        
        // if any of the aboue condition of if false, return false
        return false;
    }
};

File: Leetcode And GFG/Array/Easy/Array to BST/By Recursion.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Optimised - Recursion or Binary Tree or Divide and Conquer
// Time Complexity T - O(n)
// Space Complexity S - O(1)


class Solution {
    
    // recursive function to create BST tree
    public TreeNode buildBST(int [] nums,int startElement,int endElement){
        
        // base condition
        // ensure the start is not greater than end
        // if so, return null
        if(startElement > endElement){
            return null;
        }    
        
        // calculating the "midElement" of the given array
        int midElement = (startElement + endElement) / 2;
        
        // creating a Binary Search Tree - bst
        // by calling the recursion function (mid, left, right)
        TreeNode bst = new TreeNode(nums[midElement]);
        
        bst.left = buildBST(nums, startElement, midElement-1);
        
        bst.right = buildBST(nums, midElement+1,endElement);
        
        return bst;
    }
    
    public TreeNode sortedArrayToBST(int[] nums) {
        
        // calling recursion function
        return buildBST(nums, 0, nums.length - 1);
    }
}
